ldc-developers/ldc
# D dynamic compilation support

Here is current state of my 'jit' work. Simple examples work but everything mostly untested (and I'm not sure how properly test it). Tested on win64 and linux64. IR preparation for jit mostly in gen/runtimecompile.cpp. Jit itself implemented as shared library in runtime/rtcompile-rt/cpp-so/, it was mostly copypasted from llvm examples and I'm not sure it 100% correct.

Status:

Tests:
* I am using existing lit infrastructure, not sure it is best suitable here
* It is hard to check functions were actually dynamically compiled (if we compile them statically most of the tests still pass)
* Need more tests

Platform support:

Win64:
* Thread locals in dynamic code doesn't work
```text
Relocation type not implemented yet!
UNREACHABLE executed at c:\ldc\llvm\lib\executionengine\runtimedyld\Targets/Runt
imeDyldCOFFX86_64.h:104!
```
* Throwing and cathing exception in dynamic code doesnt work due to llvm bug (https://bugs.llvm.org//show_bug.cgi?id=24233)

Linux:
* Thread locals in dynamic code doesn't work
```text
LLVM ERROR: Cannot select: 0x1007ec8: i64 = X86ISD::WrapperRIP TargetGlobalTLSAddress:i64<i64* @_D12thread_local8threadIdm> 0 [TF=10]
  0x1007e60: i64 = TargetGlobalTLSAddress<i64* @_D12thread_local8threadIdm> 0 [TF=10]
```

MacOS:
* Thread locals doesn't work


Areas of improvement:
* Revise dynamic code optimization (copypasted from ldc/clang for now)
* Research hotpatching for dynamic code dispatch (does llvm support this?)
* Option to limit dynamic functions search depth
* Extend `@runtimeCompile` attribute to allow mark some functions always non-dynamic
* Get rid of thunks when calling dynamic functions from other dynamic functions
* Check dynamic compilation with code coverage (theoretically it should work)
* Check dynamic compilation with PGO (dynamic compiler can use PGO info collected offline)
* Research into jitted partial application, eg:
```
@rumtimeCompile int foo(int a, int b)
{
  if (b == 1)
  {
    return a * 42;
  }
  return a + 42;
}


// Generate another function which interprets 'b' as constant allow llvm optimizer to drop branch completely
auto fun = apply!foo(placeholder, 0);
return fun(5); // foo(5, 0)
```
* Throw exception when attempted to call dynamic functions without compilation
